vyper/types/convert.py

```diff
@@ -31,13 +31,21 @@
 def to_int128(expr, args, kwargs, context):
     in_node = args[0]
     typ, len = get_type(in_node)
-    if typ in ('uint256', 'bytes32'):
+    if typ is 'bytes32':
         if in_node.typ.is_literal and not SizeLimits.in_bounds('int128', in_node.value):
             raise InvalidLiteralException("Number out of range: {}".format(in_node.value), expr)
         return LLLnode.from_list(
             ['clamp', ['mload', MemoryPositions.MINNUM], in_node,
                         ['mload', MemoryPositions.MAXNUM]], typ=BaseType('int128', in_node.typ.unit), pos=getpos(expr)
```

Comment:
@jakerockland very strange that this clamp did not kick in? The limit below doesn't take the min into account.

Reply:
@jacqueswww 

Yeah it is quite strange, I spent a good while tinkering with it trying to understand _why_ that is the case that `['clamp', ['mload', MemoryPositions.MINNUM], in_node, ['mload', MemoryPositions.MAXNUM]]` catches the fail case of (2**127) but not (2**256 - 1) and was getting pretty mind boggled. 

I think I may need some direction in being pointed to where the pseudo-op codes are implemented and assembled to their composure of op-codes? Or just some direction more generally of where to look deeper into this?

I do (probably biasedly haha) think that my proposed solution here of `['uclample', in_node, SizeLimits.MAXNUM]` is a bit more elegant/readable given that we know the input is unsigned. 

Do you know the performance differences of `uclample` vs `clamp` or where I could find some reference to that/dig deeper?

Reply:
The code for clamp can be found in `https://github.com/ethereum/vyper/blob/master/vyper/compile_lll.py#L254`; and yes I agree your approach is probably better  & is less gas.

Reply:
Aha that is exactly what I was hoping to dig into in more detail. Thanks so much @jacqueswww 😄

Reply:
@jakerockland any luck figuring out why that clamp didn't work before?

Reply:
@jacqueswww Just took another look at this with fresh eyes and actually I believe so but am not 100%.

So I think the real problem with what the behavior without this PR is was not that `clamp` is broken somehow, but rather that the bounds of the `clamp` weren't correct. 

To make this clear, you can replace `['uclample', in_node, SizeLimits.MAXNUM]` on my branch, with either `['clamp', 0, in_node, ['mload', MemoryPositions.MAXNUM]]` or `['clamp', 0, in_node, SizeLimits.MAXNUM]` and the tests I added to catch the error behavior I found pass as expected.

I think clamping between the min and max of the `int128` type isn't safe when converting from `uint256` because when calling a function like this below the expression `2**256 - 1` is evaluated to be -1 and then the -1 is clamped?

```python
    code = """
@public
def test(foo: uint256) -> int128:
    return convert(foo, int128)
    """

    c = get_contract_with_gas_estimation(code)
    assert c.test(2**256 - 1) == -1
```

Does that seem like what is going on to you? 

Also one thing that I'm a bit unclear about is under what circumstances (lets say limited to conversions for now) one would want to do this `['clamp', ['mload', MemoryPositions.MINNUM], in_node, ['mload', MemoryPositions.MAXNUM]]` instead of this `['clamp', SizeLimits.MINNUM, in_node, SizeLimits.MAXNUM]`?

Given all this, for the interest of this PR it seems we _could_ switch things to use `clamp` instead of `uclample`, with the lower bound fixed, but if the latter is better for other reasons (gas usage) it seems both more clear and more efficient to use that.

Reply:
@jacqueswww Just pushed a merge conflict resolution, looking at this again with the rest of the upstream changes merged in, there are definitely other areas where the clamping going on as-is is a bit wonky/unnecessary (ex. with clamping a bool when converting from bool) which is somewhat my fault due to previous commits.

First thing I will do as part of work towards #1093 is cleaning up the clamping currently going on in type conversions. This can either be its own PR or part of WIP towards #1093.

Reply:
Maybe a side comment, but shouldn't comparisons of string values use `==` and not `is`?  I know Python can intern strings but I don't think that applies in this case.

Reply:
@jakerockland as for the initial problem, looks like we seriously have to look into fixing constant folding.